lib/puppet/resource.rb

```diff
@@ -433,12 +435,12 @@ def lookup_with_databinding(name, scope)
   private :lookup_with_databinding
 
   def lookup_in_environment(name, scope)
-    Puppet::DataProviders.lookup_in_environment(name, scope)
+    catch(:no_such_key) { Puppet::DataProviders.lookup_in_environment(name, scope, nil) }
   end
   private :lookup_in_environment
 
   def lookup_in_module(name, scope)
-    Puppet::DataProviders.lookup_in_module(name, scope)
+    catch(:no_such_key) { Puppet::DataProviders.lookup_in_module(name, scope, nil) }
```

Comment:
I didn't want the implementation to depend on a specific return value from catch. Just more to document (and public constants to share).